lib/streamlit/type_util.py

```diff
@@ -428,3 +430,31 @@ def is_version_less_than(v1: str, v2: str) -> bool:
     from packaging import version
 
     return version.parse(v1) < version.parse(v2)
+
+
+def async_generator_to_sync(
+    async_gen: AsyncGenerator[Any, Any],
+) -> Generator[Any, Any, Any]:
+    """Convert an async generator to a synchronous generator."""
+    import asyncio
+
+    close_loop = False
+    try:
+        loop = asyncio.get_running_loop()
+    except RuntimeError:
+        # No event loop is running; it is safe to create a new one
+        loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(loop)
+        close_loop = True
+
+    try:
+        # Iterate over the async generator until it raises StopAsyncIteration
+        while True:
+            yield loop.run_until_complete(async_gen.__anext__())
+    except StopAsyncIteration:
+        # The async generator has finished
+        pass
+    finally:
+        if close_loop:
+            loop.close()
+            asyncio.set_event_loop(None)
```

Comment:
Are you allowed to set this to `None` even if there was a running event loop before? I think this might mess with the "parent" process, wouldn't it?

Reply:
I added a comment to make this clear in the code